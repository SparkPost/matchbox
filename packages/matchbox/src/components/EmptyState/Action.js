import React from 'react';
import { Button } from '../Button';

const Action = React.forwardRef(function Action(props, userRef) {
  return <Button color="blue" size="large" {...props} ref={userRef} />;
});

Action.displayName = 'EmptyState.Action';
Action.propTypes = {
  ...Button.propTypes,
};
export default Action;
