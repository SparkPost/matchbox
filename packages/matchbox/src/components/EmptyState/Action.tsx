import React from 'react';
import { Button, PolymorphicButton } from '../Button';

const Action = React.forwardRef(function Action(props, userRef) {
  return <Button color="blue" size="large" {...props} ref={userRef} />;
}) as PolymorphicButton;

Action.displayName = 'EmptyState.Action';
export default Action;
