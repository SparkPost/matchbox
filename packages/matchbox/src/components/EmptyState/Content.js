import React from 'react';
import PropTypes from 'prop-types';
import { StyledContent } from './styles';

function Content(props) {
  const { children } = props;

  return <StyledContent>{children}</StyledContent>;
}

Content.displayName = 'EmptyState.Content';

Content.propTypes = {
  children: PropTypes.node,
};

export default Content;
