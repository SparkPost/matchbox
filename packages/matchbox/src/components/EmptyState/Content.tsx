import React from 'react';
import { StyledContent } from './styles';

type EmptyStateContentProps = {
  children?: React.ReactNode;
};

const Content = React.forwardRef<HTMLDivElement, EmptyStateContentProps>(function Content(
  props,
  userRef,
) {
  const { children } = props;

  return <StyledContent ref={userRef}>{children}</StyledContent>;
});

Content.displayName = 'EmptyState.Content';
export default Content;
