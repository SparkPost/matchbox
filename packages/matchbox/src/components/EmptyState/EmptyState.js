import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { buttonFrom } from '../Button';
import { linkFrom } from '../UnstyledLink';
import styles from './EmptyState.module.scss';

class EmptyState extends Component {
  static displayName = 'EmptyState';

  static propTypes = {
    /**
     * The display title
     */
    title: PropTypes.string,
    /**
      * Main cta. Most button props will work in here.
      * e.g. { content: 'button label', onClick: callback() }
      */
    primaryAction: PropTypes.shape({
      content: PropTypes.node.isRequired
    }),
    /**
      * Secondary Action - appear as a link next to the primary action
      */
    secondaryAction: PropTypes.shape({
      content: PropTypes.node.isRequired
    }),
    /**
      * Image to display
      */
    image: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.element
    ]),
    /**
      * Content below the CTA
      */
    children: PropTypes.node
  };

  render() {
    const {
      title,
      primaryAction,
      secondaryAction,
      image: Image,
      children
    } = this.props;

    const imageMarkup = Image
      ? <div className={styles.Image}><Image /></div>
      : null;

    const primaryActionMarkup = primaryAction
      ? buttonFrom(primaryAction, {
        size: 'large',
        ...(!primaryAction.color ? { color: 'orange' } : {})
      })
      : null;

    const secondaryActionMarkup = secondaryAction
      ? <span className={styles.SecondaryAction}>{linkFrom(secondaryAction)}</span>
      : null;

    return (
      <div className={styles.EmptyState}>
        <h1 className={styles.Title}>{title}</h1>
        <div className={styles.Content}>{children}</div>
        <div className={styles.Actions}>
          {primaryActionMarkup} {secondaryActionMarkup}
        </div>
        {imageMarkup}
      </div>
    );
  }
}

export default EmptyState;
