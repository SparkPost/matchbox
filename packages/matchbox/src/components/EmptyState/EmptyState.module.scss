@import '../../styles/config';

.EmptyState {
  position: relative;
  height: 100vh;
  padding: 33vh spacing();
}

.Title {
  @include text-style-display-heading;
  margin-bottom: spacing();
  width: auto;

  @media screen and (min-width: breakpoint(small)) {
    width: 45%;
  }
}

.Content {
  width: auto;

  @media screen and (min-width: breakpoint(small)) {
    width: 45%;
  }

  p {
    font-size: font-size(600);
    color: color(gray, 3);
    margin-bottom: spacing();
  }
}

.Actions {
  width: auto;

  @media screen and (min-width: breakpoint(small)) {
    width: 45%;
  }

  & > * {
    margin-right: spacing();
  }
}

.SecondaryAction {
  display: inline-block;
  padding-top: spacing();
  white-space: nowrap;
}

.Image {
  display: none;
  position: absolute;
  top: 45%;
  left: 45%;
  width: 50%;
  max-width: rem(600);
  height: auto;

  transform: translate(0, -50%);

  & > svg {
    height: 100%;
    width: 100%;
  }

  @media screen and (min-width: breakpoint(small)) {
    display: block;
  }
}
