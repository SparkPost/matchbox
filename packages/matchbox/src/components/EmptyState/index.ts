export { default as EmptyState } from './EmptyState';
export * from './Content';
export * from './EmptyState';
export * from './Header';
export * from './Image';
export * from './List';
export * from './Media';
