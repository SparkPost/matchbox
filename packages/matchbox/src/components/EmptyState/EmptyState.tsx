import React from 'react';
import { margin, MarginProps } from 'styled-system';
import { Box } from '../Box';
import { Columns } from '../Columns';
import { Column } from '../Column';
import { Inline } from '../Inline';
import Action from './Action';
import Header from './Header';
import Content from './Content';
import List from './List';
import Image from './Image';
import Media from './Media';
import Legacy from './Legacy';
import { getChild } from '../../helpers/children';
import { pick } from '../../helpers/props';
import useBreakpoint from '../../hooks/useBreakpoint';

export type EmptyStateProps = {
  children?: React.ReactNode;
} & MarginProps;

const EmptyState = React.forwardRef<HTMLDivElement, EmptyStateProps>(function EmptyState(
  props,
  userRef,
) {
  const { children, ...rest } = props;
  const breakpoint = useBreakpoint();
  const isMobile = ['default', 'xs', 'sm'].includes(breakpoint);

  const systemProps = pick(rest, margin.propNames);

  // Allow fallback use of `EmptyState.Image` for now
  // We should remove in the future in favor of `EmptyState.Media`
  const Media = getChild('EmptyState.Media', children);
  const Image = getChild('EmptyState.Image', children);

  return (
    <Columns collapseBelow="sm" space="500" alignY="center" {...systemProps} ref={userRef}>
      {isMobile ? (
        <Column width={1 / 2}>{Image && Image.length >= 1 ? Image : Media}</Column>
      ) : null}
      <Column width={1 / 2}>
        {getChild('EmptyState.Header', children)}
        {getChild('EmptyState.Content', children)}
        <Box mt={650}>
          <Inline space="500">{getChild('EmptyState.Action', children)}</Inline>
        </Box>
      </Column>
      {!isMobile ? (
        <Column width={1 / 2}>{Image && Image.length >= 1 ? Image : Media}</Column>
      ) : null}
    </Columns>
  );
}) as React.ForwardRefExoticComponent<EmptyStateProps> & {
  Action: typeof Action;
  Header: typeof Header;
  Content: typeof Content;
  Image: typeof Image;
  Media: typeof Media;
  List: typeof List;
  LEGACY: typeof Legacy;
};

EmptyState.displayName = 'EmptyState';
EmptyState.Action = Action;
EmptyState.Header = Header;
EmptyState.Content = Content;
EmptyState.Image = Image;
EmptyState.Media = Media;
EmptyState.List = List;
EmptyState.LEGACY = Legacy;

export default EmptyState;
