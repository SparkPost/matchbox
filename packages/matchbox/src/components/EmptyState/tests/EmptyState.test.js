import React from 'react';
import EmptyState from '../EmptyState';
import { shallow } from 'enzyme';

describe('EmptyState', () => {
  let wrapper;
  let props;
  beforeEach(() => {
    props = {
      title: 'Not there yet!'
    };

    wrapper = shallow(<EmptyState {...props}><p>Nothing here yet!</p></EmptyState>);
  });


  it('renders with image', () => {
    wrapper.setProps({ image: jest.fn() });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without any actions', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with actions', () => {
    wrapper.setProps({
      primaryAction: { content: 'button', onClick: jest.fn() },
      secondaryAction: { content: 'secondary', onClick: jest.fn() }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with primary action', () => {
    wrapper.setProps({ primaryAction: { content: 'button', onClick: jest.fn() }});
    expect(wrapper).toMatchSnapshot();
  });

});
