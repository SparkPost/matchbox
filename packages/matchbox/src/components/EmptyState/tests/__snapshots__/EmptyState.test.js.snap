// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyState renders with actions 1`] = `
<div
  className="EmptyState"
>
  <h1
    className="Title"
  >
    Not there yet!
  </h1>
  <div
    className="Content"
  >
    <p>
      Nothing here yet!
    </p>
  </div>
  <div
    className="Actions"
  >
    <Button
      color="orange"
      onClick={[MockFunction]}
      size="large"
    >
      button
    </Button>
     
    <span
      className="SecondaryAction"
    >
      <UnstyledLink
        onClick={[MockFunction]}
      >
        secondary
      </UnstyledLink>
    </span>
  </div>
</div>
`;

exports[`EmptyState renders with image 1`] = `
<div
  className="EmptyState"
>
  <h1
    className="Title"
  >
    Not there yet!
  </h1>
  <div
    className="Content"
  >
    <p>
      Nothing here yet!
    </p>
  </div>
  <div
    className="Actions"
  >
     
  </div>
  <div
    className="Image"
  >
    <mockConstructor />
  </div>
</div>
`;

exports[`EmptyState renders with primary action 1`] = `
<div
  className="EmptyState"
>
  <h1
    className="Title"
  >
    Not there yet!
  </h1>
  <div
    className="Content"
  >
    <p>
      Nothing here yet!
    </p>
  </div>
  <div
    className="Actions"
  >
    <Button
      color="orange"
      onClick={[MockFunction]}
      size="large"
    >
      button
    </Button>
     
  </div>
</div>
`;

exports[`EmptyState renders without any actions 1`] = `
<div
  className="EmptyState"
>
  <h1
    className="Title"
  >
    Not there yet!
  </h1>
  <div
    className="Content"
  >
    <p>
      Nothing here yet!
    </p>
  </div>
  <div
    className="Actions"
  >
     
  </div>
</div>
`;
